pkg/settlement/swap/transaction/monitor.go

```diff
@@ -0,0 +1,251 @@
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	t := &transactionMonitor{
+		ctx:        ctx,
+		cancelFunc: cancelFunc,
+		logger:     logger,
+		backend:    backend,
+		sender:     sender,
+
+		pollingInterval:   pollingInterval,
+		cancellationDepth: cancellationDepth,
+
+		watches:    make(map[*transactionWatch]struct{}),
+		watchAdded: make(chan struct{}, 1),
+	}
+
+	t.wg.Add(1)
+	go t.watchPending()
+
+	return t
+}
+
+func (tm *transactionMonitor) WatchTransaction(txHash common.Hash, nonce uint64) (chan types.Receipt, chan error, error) {
+	tm.lock.Lock()
+	defer tm.lock.Unlock()
+
+	// these channels will be written to at most once
+	// buffer size is 1 to avoid blocking in the watch loop
+	receiptC := make(chan types.Receipt, 1)
+	errC := make(chan error, 1)
+
+	tm.watches[&transactionWatch{
+		receiptC: receiptC,
+		errC:     errC,
+		txHash:   txHash,
+		nonce:    nonce,
+	}] = struct{}{}
+
+	select {
+	case tm.watchAdded <- struct{}{}:
+	default:
+	}
+
+	tm.logger.Tracef("starting to watch transaction %x with nonce %d", txHash, nonce)
+
+	return receiptC, errC, nil
+}
+
+// main watch loop
+func (tm *transactionMonitor) watchPending() {
+	defer tm.wg.Done()
+	defer func() {
+		tm.lock.Lock()
+		defer tm.lock.Unlock()
+
+		for watch := range tm.watches {
+			watch.errC <- ErrMonitorClosed
+		}
+	}()
+
+	var lastBlock uint64 = 0
+
+	for {
+		var added bool = false // flag if this iteration was triggered by the watchAdded channel
+		select {
+		// if a new watch has been added check again without waiting
+		case <-tm.watchAdded:
+			added = true
+		// otherwise wait
+		case <-time.After(tm.pollingInterval):
+		// if the main context is cancelled terminate
+		case <-tm.ctx.Done():
+			return
+		}
+
+		// if there are no watched transactions there is nothing to do
+		if !tm.hasWatches() {
+			continue
+		}
+
+		// switch to new head subscriptions once websockets are the norm
+		block, err := tm.backend.BlockNumber(tm.ctx)
+		if err != nil {
+			tm.logger.Errorf("could not get block number: %v", err)
+			continue
+		} else if block <= lastBlock && !added {
+			// if the block number is not higher than before there is nothing todo
+			// unless a watch was added in which case we will do the check anyway
+			// in the rare case where a block was reorged and the new one is the first to contain our tx we wait an extra block
+			continue
+		}
+
+		if err := tm.checkPending(block); err != nil {
+			tm.logger.Errorf("error while checking pending transactions: %v", err)
+		}
+
+		lastBlock = block
+	}
+}
+
+type confirmedTx struct {
+	receipt types.Receipt
+	watch   *transactionWatch
+}
+
+// potentiallyConfirmedWatches returns all watches with nonce less than what was specified
+func (tm *transactionMonitor) potentiallyConfirmedWatches(nonce uint64) (watches []*transactionWatch) {
+	tm.lock.Lock()
+	defer tm.lock.Unlock()
+
+	for watch := range tm.watches {
```

Comment:
this code block sent me thinking quite a lot and writing quite a few code examples in the playground to see that the for iterator here won't pass the same allocated pointer. it seems to be ok, but consider that others might spend time understanding whether this works as expected or not in the future...